In [1]:
import pandas as pd 

In [2]:
df = pd.read_excel(r"D:\Project\online+retail\Online Retail.xlsx")

In [3]:
# display the first 5 rows
df.head() 

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [4]:
# display the number of rows and columns 
df.shape

(541909, 8)

In [5]:
# display all column names 
df.columns

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [6]:
# shows the data types and missing-value information
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


### Reprecentation of one each row 
- each row represents one product line within a incoive.
One invoice can contain multiple products, so one invoice can appear in multiple rows. 

### Dataset size 
- number of row: 541909
- number of columns: 8 

### Data Dictionary
- InvoiceNo (object): Unique idetifier for an invoice or transaction
- StockCode (object): Unique identifier for a product 
- Description (object): Name or decription of the product 
- Quantity (int64): Number of product units in the tansaction
- InvoiceData (detertime64): data and time of the transaction
- UniPrice (float64): Price of one product unit
- CustomerID (float64): Identifier for the customer
- Country (object): Country where the customer is located  

In [7]:
import numpy as np

In [8]:
# check for missing values 
print(df.isna().sum())

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64


In [9]:
# calcuate the percentage of missing value 
print((df.isna().sum()/len(df))*100)

InvoiceNo       0.000000
StockCode       0.000000
Description     0.268311
Quantity        0.000000
InvoiceDate     0.000000
UnitPrice       0.000000
CustomerID     24.926694
Country         0.000000
dtype: float64


In [10]:
missing_summary = pd.DataFrame({
    "Missing_count": df.isna().sum(),
    "Missing_percentage": (df.isna().sum() / len(df)) * 100
})

missing_summary

,Missing_count,Missing_percentage
InvoiceNo,0,0.000000
StockCode,0,0.000000
Description,1454,0.268311
Quantity,0,0.000000
InvoiceDate,0,0.000000
UnitPrice,0,0.000000
CustomerID,135080,24.926694
Country,0,0.000000


In [11]:
df.duplicated().sum()

np.int64(5268)

In [12]:
# check for duplicate rows 
duplicate_rows = df[df.duplicated(keep=False)]
duplicate_rows.head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
485,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
489,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
494,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
521,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908.0,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908.0,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
548,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920.0,United Kingdom
555,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920.0,United Kingdom


In [13]:
duplicate_rows.sort_values(
    by=["InvoiceNo", "StockCode", "InvoiceDate"]
).head(20)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
494,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
485,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
489,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
521,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908.0,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908.0,United Kingdom
565,536412,21448,12 DAISY PEGS IN WOOD BOX,2,2010-12-01 11:49:00,1.65,17920.0,United Kingdom
578,536412,21448,12 DAISY PEGS IN WOOD BOX,1,2010-12-01 11:49:00,1.65,17920.0,United Kingdom


In [14]:
duplicate_percentage = df.duplicated().sum() / len(df) * 100
duplicate_percentage

np.float64(0.9721189350979592)

In [15]:
# Negative quantities (cancellation or return)
negative_quantity = (df["Quantity"] < 0).sum()
negative_quantity

np.int64(10624)

In [16]:
df[df['Quantity']<0].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
238,C536391,21980,PACK OF 12 RED RETROSPOT TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
239,C536391,21484,CHICK GREY HOT WATER BOTTLE,-12,2010-12-01 10:24:00,3.45,17548.0,United Kingdom
240,C536391,22557,PLASTERS IN TIN VINTAGE PAISLEY,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
241,C536391,22553,PLASTERS IN TIN SKULLS,-24,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
939,C536506,22960,JAM MAKING SET WITH JARS,-6,2010-12-01 12:38:00,4.25,17897.0,United Kingdom


In [17]:
# Zero quantities
zero_quantity = (df['Quantity']==0).sum()
zero_quantity

np.int64(0)

In [18]:
# Negative Prices 
negative_price=(df['UnitPrice']<0).sum()
negative_price

np.int64(2)

In [19]:
df[df["UnitPrice"] < 0].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom


In [20]:
#zero price = free product 
zero_price = (df["UnitPrice"] == 0).sum()
zero_price

np.int64(2515)

In [21]:
df[df["UnitPrice"] == 0].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
1988,536550,85044,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
2024,536552,20950,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
2025,536553,37461,NaN,3,2010-12-01 14:35:00,0.0,NaN,United Kingdom
2026,536554,84670,NaN,23,2010-12-01 14:35:00,0.0,NaN,United Kingdom
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,NaN,United Kingdom


In [22]:
# cancellation 
cancelled_invoice = (
    df["InvoiceNo"]
    .astype(str)
    .str.startswith("C")
    .sum()
)

cancelled_invoice

np.int64(9288)

In [23]:
df[
    df["InvoiceNo"].astype(str).str.startswith("C")
].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,14527.0,United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,15311.0,United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
238,C536391,21980,PACK OF 12 RED RETROSPOT TISSUES,-24,2010-12-01 10:24:00,0.29,17548.0,United Kingdom
239,C536391,21484,CHICK GREY HOT WATER BOTTLE,-12,2010-12-01 10:24:00,3.45,17548.0,United Kingdom
240,C536391,22557,PLASTERS IN TIN VINTAGE PAISLEY,-12,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
241,C536391,22553,PLASTERS IN TIN SKULLS,-24,2010-12-01 10:24:00,1.65,17548.0,United Kingdom
939,C536506,22960,JAM MAKING SET WITH JARS,-6,2010-12-01 12:38:00,4.25,17897.0,United Kingdom


In [24]:
quality_counts = pd.Series({
    "Exact duplicate rows": df.duplicated().sum(),
    "Negative quantity rows": negative_quantity,
    "Zero quantity rows": zero_quantity,
    "Negative price rows": negative_price,
    "Zero price rows": zero_price,
    "Cancelled invoice rows": cancelled_invoice
})

quality_counts

Exact duplicate rows       5268
Negative quantity rows    10624
Zero quantity rows            0
Negative price rows           2
Zero price rows            2515
Cancelled invoice rows     9288
dtype: int64

- Duplicate rows: what does it mean?
- Negative quantity rows: retruned products, cancelled orders, damaged-product, incentory corrections...
- Zero quantity: there are no row for this 
- Negative price: data errors, accounting adjustments 
- Zero price rows: free staffs, prometional products, missing price info, data entry problems 
- Cancelled invoice 

### Assume that relationship between negative quantity and cancelled invoice

- Negative quantity rows are 10624 
- Cancelled invoice rows are 9288

- Assume that every cancelled row has a negative quantity
(10624-2988 = 1336) -> 1336 negatice quantity rows do not begin with C

In [25]:
# check for the relationship between negative quantity and cancelled invoice
cancelled_mask = df["InvoiceNo"].astype(str).str.startswith("C")
negative_quantity_mask = df["Quantity"] <0

(cancelled_mask & negative_quantity_mask).sum()

np.int64(9288)

In [26]:
(negative_quantity_mask & ~cancelled_mask).sum()

np.int64(1336)

In [27]:
df[negative_quantity_mask & ~cancelled_mask].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,NaN,United Kingdom
4347,536764,84952C,NaN,-38,2010-12-02 14:42:00,0.0,NaN,United Kingdom
7188,536996,22712,NaN,-20,2010-12-03 15:30:00,0.0,NaN,United Kingdom
7189,536997,22028,NaN,-20,2010-12-03 15:30:00,0.0,NaN,United Kingdom
7190,536998,85067,NaN,-6,2010-12-03 15:30:00,0.0,NaN,United Kingdom
7192,537000,21414,NaN,-22,2010-12-03 15:32:00,0.0,NaN,United Kingdom
7193,537001,21653,NaN,-6,2010-12-03 15:33:00,0.0,NaN,United Kingdom
7195,537003,85126,NaN,-2,2010-12-03 15:33:00,0.0,NaN,United Kingdom
7196,537004,21814,NaN,-30,2010-12-03 15:34:00,0.0,NaN,United Kingdom
7197,537005,21692,NaN,-70,2010-12-03 15:35:00,0.0,NaN,United Kingdom


So all of cancellation invioce has negative quantity

# Clean the data

In [28]:
# copy of the original data
clean_df = df.copy()

In [29]:
# starting rows 
starting_rows = len(clean_df)
starting_rows

541909

In [30]:
# create a cleaning log for recording 
cleaning_log = []

cleaning_log.append({
    "Step":"Initial data",
    "Rows_Before":len(clean_df),
    "Rows_After": len(clean_df),
    "Rows_Removed": 0
})

In [31]:
# Remove exact duplicates
rows_before_du = len(clean_df)

In [32]:
clean_df = clean_df.drop_duplicates().copy()

In [33]:
rows_after_du = len(clean_df)

In [34]:
rows_removed_du = (rows_before_du - rows_after_du)

rows_removed_du

5268

In [35]:
cleaning_log.append({
    "Step":"Removed exact duplicates",
    "Rows_Before":rows_before_du,
    "Rows_After": rows_after_du,
    "Rows_Removed": rows_removed_du
})

In [36]:
cleaning_log

[{'Step': 'Initial data',
  'Rows_Before': 541909,
  'Rows_After': 541909,
  'Rows_Removed': 0},
 {'Step': 'Removed exact duplicates',
  'Rows_Before': 541909,
  'Rows_After': 536641,
  'Rows_Removed': 5268}]

In [37]:
clean_df.shape

(536641, 8)

In [38]:
#convert the InvoiceNo to string (text)
clean_df["InvoiceNo"]=(clean_df["InvoiceNo"].astype("string"))

In [39]:
#convert the StockCode to string (text)
clean_df["StockCode"]=(clean_df["StockCode"].astype("string"))

In [40]:
# Convert the CustomerID from decimal to integer, then text
clean_df["CustomerID"] = (
    clean_df["CustomerID"].astype("Int64").astype("string")
)

In [41]:
# Ensure the data is a datetime 
clean_df["InvoiceDate"] = pd.to_datetime(
    clean_df["InvoiceDate"],
    errors="coerce"
)

In [42]:
#cehck the data types of the columns
clean_df.dtypes

InvoiceNo      string[python]
StockCode      string[python]
Description            object
Quantity                int64
InvoiceDate    datetime64[ns]
UnitPrice             float64
CustomerID     string[python]
Country                object
dtype: object

In [43]:
# organise the space in the "description" column
clean_df["Description"] = (
    clean_df["Description"]
    .astype("string")
    .str.replace(r"\s+", " ", regex=True)
    .str.strip()
)

In [44]:
# organise the space in the "country" column
clean_df["Country"] = (
    clean_df["Country"]
    .astype("string")
    .str.replace(r"\s+", " ", regex=True)
    .str.strip()
)

## Create transaction status 

In [45]:
clean_df["TransactionStatus"] = "Completed Sales"

In [47]:
# identify missing values in the InvoiceNo column
cancelled_mask = (
    clean_df["InvoiceNo"].str.startswith("C", na=False)
)

In [48]:
# identify cancelled transactions for updating the TransactionStatus column
clean_df.loc[
    cancelled_mask,
    "TransactionStatus"
] = "Cancelled"

In [49]:
# identify negative quantities without C
stock_adjustment_mask = (
    (clean_df["Quantity"]<0) & ~cancelled_mask
)

In [50]:
# identify stock adjustment transactions for updating the TransactionStatus column
clean_df.loc[
    stock_adjustment_mask,
    "TransactionStatus"
] = "Stock Adjustment"

In [51]:
#identify zero price rows 
zero_price_mask = clean_df["UnitPrice"] == 0

In [52]:
#identify zero price transactions for updating the TransactionStatus column
clean_df.loc[
    zero_price_mask,
    "TransactionStatus"
] = "Price Adjustment"

In [53]:
#identify negative price rows
negative_price_mask = clean_df["UnitPrice"] < 0

In [54]:
# identify negative price transactions for updating the TransactionStatus column
clean_df.loc[
    negative_price_mask,
    "TransactionStatus"
] = "Price Adjustment"

In [55]:
# cehck the TransactionStatus 
clean_df["TransactionStatus"].value_counts(
    dropna=False
)

TransactionStatus
Completed Sales     524878
Cancelled             9251
Price Adjustment      2512
Name: count, dtype: int64

In [56]:
# check the examples 
clean_df[
    [
        "InvoiceNo",
        "Description",
        "Quantity",
        "UnitPrice",
        "TransactionStatus"
    ]
].head(20)

,InvoiceNo,Description,Quantity,UnitPrice,TransactionStatus
0,536365,WHITE HANGING HEART T-LIGHT HOLDER,6,2.55,Completed Sales
1,536365,WHITE METAL LANTERN,6,3.39,Completed Sales
2,536365,CREAM CUPID HEARTS COAT HANGER,8,2.75,Completed Sales
3,536365,KNITTED UNION FLAG HOT WATER BOTTLE,6,3.39,Completed Sales
4,536365,RED WOOLLY HOTTIE WHITE HEART.,6,3.39,Completed Sales
5,536365,SET 7 BABUSHKA NESTING BOXES,2,7.65,Completed Sales
6,536365,GLASS STAR FROSTED T-LIGHT HOLDER,6,4.25,Completed Sales
7,536366,HAND WARMER UNION JACK,6,1.85,Completed Sales
8,536366,HAND WARMER RED POLKA DOT,6,1.85,Completed Sales
9,536367,ASSORTED COLOUR BIRD ORNAMENT,32,1.69,Completed Sales


In [57]:
cleaning_log.append({
    "Step": "Corrected types, standardised text and labelled status",
    "Rows_Before": len(clean_df),
    "Rows_After": len(clean_df),
    "Rows_Removed": 0
})

In [58]:
cleaning_log_df = pd.DataFrame(cleaning_log)

cleaning_log_df

,Step,Rows_Before,Rows_After,Rows_Removed
0,Initial data,541909,541909,0
1,Removed exact duplicates,541909,536641,5268
2,"Corrected types, standardised text and labelle...",536641,536641,0


In [59]:
clean_df.shape

(536641, 9)

In [60]:
clean_df.dtypes

InvoiceNo            string[python]
StockCode            string[python]
Description          string[python]
Quantity                      int64
InvoiceDate          datetime64[ns]
UnitPrice                   float64
CustomerID           string[python]
Country              string[python]
TransactionStatus            object
dtype: object

In [61]:
clean_df.isna().sum()

InvoiceNo                 0
StockCode                 0
Description            1454
Quantity                  0
InvoiceDate               0
UnitPrice                 0
CustomerID           135037
Country                   0
TransactionStatus         0
dtype: int64

In [62]:
clean_df.duplicated().sum()

np.int64(0)

In [63]:
clean_df["TransactionStatus"].value_counts()

TransactionStatus
Completed Sales     524878
Cancelled             9251
Price Adjustment      2512
Name: count, dtype: int64